### ISSUE TITLE
Refund transaction returns error code E0027, no details

### ISSUE THREAD

**Issue by dcartwright** [AuthorizeNet/sdk-dotnet#80]
I am testing code to perform a refund transaction, and it is failing with this same E0027 error as described here: https://github.com/AuthorizeNet/sdk-dotnet/issues/73.

I am using a sandbox account and the "Controller" method for accessing the API. I am able to submit test AuthorizeAndCapture charges and Void them, getting correct responses. However, when I go to a refundTransaction against an existing unsettled charge, I am getting this E0027 error with absolutely no additional detail: all fields of transactionResponse are null, and there is no text description on the message.

In the issue I linked to, the suspected cause of the problem is an AVS validation error. However, for testing purposes I tried disabling both AVS and CCV filtering entirely on my sandbox account, and the error persists.

Here is a version of the code I am using, edited for clarity:

```
var request = new createTransactionRequest {
    transactionRequest = new transactionRequestType {
        transactionType = transactionTypeEnum.refundTransaction.ToString(),
        refTransId = transaction.TransactionID,
        amount = amount.Value,
        profile = new customerProfilePaymentType {
            customerProfileId = transaction.CreditCardProfile.Profile.AuthNetCustomerID,
            paymentProfile = new paymentProfile {
                paymentProfileId = transaction.CreditCardProfile.AuthNetPaymentID,
                cardCode = transaction.CreditCardProfile.CCV,
            }
        },
    }
};
var response = MakeRequest( request );
```

The MakeRequest() method shown ultimately does something like this:
```
var controller = new transactionRequestController( request );
var response = controller.ExecuteWithApiResponse( Environment );
```

Then the following statements hold true for the returned response:
`response.transactionResponse.* == null`
`response.messages.resultCode == messageTypeEnum.Error`
`response.messages.message[0].code == "E0027"`
`response.messages.message[0].text == null`

**Comment by ramittal** [comment]
Since the error originated on the transctionProcessing, the error details are embedded inside the transactionResponse...You can see the error details at creditResp.transactionResponse.errors[0].

**Comment by brianmc** [comment]
Hi @dcartwright, you can see a sample refund/credit transaction here:  https://github.com/AuthorizeNet/sdk-dotnet/blob/future/AuthorizeNETtest/Api/Controllers/SampleTest/CreateTransactionSampleTest.cs

You can not refund a transaction until it is settled. For unsettled transactions, a Void is the equivalent request required.   But you bring up a common developer request:  "we don't care if it's a refund or a void, we just want to cancel the transaction".

**Comment by brianmc** [comment]
You should get a transaction response error 54:

<error>
        <errorCode>
          54
        </errorCode>
        <errorText>
          The referenced transaction does not meet the criteria for issuing a credit.
        </errorText>
</error>

You can always test it out here as well : http://developer.authorize.net/api/reference/#payment-transactions-refund-a-transaction

**Comment by dcartwright** [comment]
@brianmc  - Yes, thank you, I think you are probably right that the conceptual mistake of Refund-before-settled is probably what is causing my problem. That's why I can charge and then immediately void, but the refund fails.

However, there is still a bug. As @ramittal points out, I should be able to see a transactionResponse.errors collection and, as you say, get an error code 54. However, I do not - instead transactionResponse.errors is null, and I get an error code 27 (which is supposedly an AVS validation error), over in the generic response messages. I imagine that the underlying bug here is the same as #73.

**Comment by brianmc** [comment]
Apologies @dcartwright I was focused on the refund/void aspect rather than the fact that you were using profiles.  Yes indeed #73 is indicative of an underlying API issue with error responses from createTransaction when using profiles.

Rest assured we have the fix for this and it is in the pipeline to get released.  I'm reopening this issue until that happens.

Brian

**Comment by brianmc** [comment]
Closed since we are using #73 to track this.  Watch that bug, it'll stay open until we fix the issue.
